torchacc/dist/fsdp.py

```diff
@@ -202,3 +206,179 @@ def clip_grad_norm_(self, max_grad_norm):
+            for fn in flatten_name_list:
+                consolidate_optim_state_dict['param_groups'][0][
+                    'params'].append(fn)
+
+        unflat_state_dict = {fn: {} for fn in flatten_name_list}
+
+        for idx, layer_state in enumerate(sharded_optim_state.values()):
+            layer_names = layer_name_lists[idx]
+            layer_shapes = layer_size_lists[idx]
+            layer_numels = layer_numel_lists[idx]
+            for state_name, state_params in layer_state.items():
+                tensor_buffer = optim_utils.all_gather_state(
+                    state_params, self.model)
+                orig_params = optim_utils.unflatten_optim_params(
+                    tensor_buffer, layer_names, layer_shapes, layer_numels)
+
+                if not rank0_only or self.model.rank == 0:
+                    for fn, fp in zip(layer_names, orig_params):
+                        if cpu_offload:
+                            xm.mark_step()
+                            unflat_state_dict[fn][state_name] = fp.cpu()
+                        else:
+                            unflat_state_dict[fn][state_name] = fp
+                xm.mark_step()
+        consolidate_optim_state_dict['state'] = unflat_state_dict
+
+        return consolidate_optim_state_dict
+
+    def load_optim_state_dict(self,
+                              optim_state_dict: Dict[str, Any],
+                              rank0_only: bool = True) -> Dict[str, Any]:
+        """
+        Convert an optimizer state-dict so that it can be loaded into the
+        optimizer associated with the FSDP model.
+        We judge whether the optim_state_dict is sharded automatically
+                
+        Args:
+            optim_state_dict (Dict[str, Any]): The optimizer states to be loaded.
+            rank0_only: (bool): control whether load state_dict only from
+                rank0 at the begining.(Default: ``True``)
+        
+        Returns:
+            Dict[str, Any]: A :class:`dict` containing the optimizer state for
+            self.model which is sharded.
+        """
+        # for sharded optim_state, we return directly
+        if 'shard_metadata' in optim_state_dict.keys():
+            return optim_state_dict['optimizer']
+
+        shard_meta_data = self.model.get_shard_metadata()
+        unflat_optim_state = optim_state_dict
+        flat_optim_state: Dict[str, Any] = {'state': {}, 'param_groups': {}}
+
+        layer_name_lists, layer_size_lists, layer_numel_lists = optim_utils.get_layer_full_info(
+            shard_meta_data, self.state_dict())
+
+        if rank0_only:
+            unflat_optim_state = optim_utils.broadcast_processed_state(
+                unflat_optim_state, self.model.rank, self.model.sharding_groups)
+        unflat_state = unflat_optim_state['state']
+
+        flat_optim_state['param_groups'] = copy.deepcopy(
+            unflat_optim_state['param_groups'])
+        # flatten and sharded state_dict
```

Comment:
we may need to check model's `flatten_parameters` and raise error if not

Reply:
sorry the comments I made may cause confusion, the implementation below is work for all scenarios and I have removed the comments.